simple_parsing/replace.py

```diff
@@ -90,3 +104,80 @@ def replace(obj: DataclassT, changes_dict: dict[str, Any] | None = None, **chang
     replace_kwargs.update(changes)
 
     return dataclasses.replace(obj, **replace_kwargs)
+
+
+def unflatten_selection_dict(
+    flattened: Mapping[str, V], keyword: str = "__key__", sep="."
+) -> PossiblyNestedDict[str, V]:
+    """
+    This function convert a flattened dict into a nested dict
+    and it inserts the `keyword` as the selection into the nested dict.
+
+    >>> unflatten_selection_dict({'ab_or_cd': 'cd', 'ab_or_cd.c_or_d': 'd'})
+    {'ab_or_cd': {'__key__': 'cd', 'c_or_d': 'd'}}
+
+    >>> unflatten_selection_dict({"a": 1, "b": 2})
+    {'a': {'__key__': 1}, 'b': {'__key__': 2}}
+    """
+    dc = {}
+    for k, v in flattened.items():
+        if keyword != k and sep not in k and not isinstance(v, dict):
+            dc[k + sep + keyword] = v
+        else:
+            dc[k] = v
+    return unflatten_split(dc)
+
+
+def replace_subgroups(obj: DataclassT, selections: dict[str, Key | DataclassT] | None = None):
+    """
+    This function replaces the dataclass of subgroups, union, and optional union.
+    The `selections` dict can be in flat format or in nested format.
+
+    The values of selections can be `Key` of subgroups, dataclass type, and dataclass instance.
+    """
+    keyword = "__key__"
+
+    if selections:
+        selections = unflatten_selection_dict(selections, keyword)
+    else:
+        return obj
+
+    replace_kwargs = {}
+    for field in dataclasses.fields(obj):
+        if field.name not in selections:
+            continue
+
+        field_value = getattr(obj, field.name)
+        t = get_field_type_from_annotations(obj.__class__, field.name)
+
+        new_value = None
+        # Replace subgroup is allowed when the type annotation contains dataclass
+        if contains_dataclass_type_arg(t):
+            child_selections = selections.pop(field.name)
+            key = child_selections.pop(keyword, None)
+
+            if is_dataclass_type(key):
+                field_value = key()
+            elif is_dataclass_instance(key):
+                field_value = copy.deepcopy(key)
+            elif field.metadata.get("subgroups", None):
+                field_value = field.metadata["subgroups"][key]()
+            elif is_optional(t) and key is None:
+                field_value = None
+            elif contains_dataclass_type_arg(t) and key is None:
+                field_value = field.default_factory()
+            else:
+                raise ValueError(f"invalid selection key '{key}' for field '{field.name}'")
+
+            if child_selections:
+                new_value = replace_subgroups(field_value, child_selections)
+            else:
+                new_value = field_value
+        else:
+            raise ValueError(f"The replaced subgroups contains no dataclass in its annotation {t}")
+
+        if not field.init:
+            raise ValueError(f"Cannot replace value of non-init field {field.name}.")
```

Comment:
I suggest moving this before all the code above.

Reply:
Will fix this.